cunumeric/vectorize.py

```diff
@@ -0,0 +1,643 @@
+        self._numba_func: Callable[[Any], Any]
+        self._cpu_func: numba.core.ccallback.CFunc
+        self._gpu_func: tuple[Any]
+        self._args: List[Any] = []
+        self._scalar_args: List[Any] = []
+        self._scalar_idxs: List[int] = []
+        self._scalar_names: List[str] = []
+        self._arg_names: List[str] = []
+        self._kwargs: List[Any] = []
+        self._context = runtime.legate_context
+        self._created: bool = False
+        self._func_body: List[str] = []
+
+        if doc is None:
+            self.__doc__ = pyfunc.__doc__
+        else:
+            self.__doc__ = doc
+
+        self._return_names = self._get_return_argumets()
+        self._num_outputs: int = len(self._return_names)
+        self._return_args: List[Any] = []
+        self._output_dtype: Optional[np.dtype[Any]] = None
+        self._cached_dtype: Optional[np.dtype[Any]] = None
+        self._cached_scalar_types: List[Any] = []
+
+        if otypes is not None:
+            if self._num_outputs != len(otypes):
+                raise ValueError(
+                    "number of types in otypes is not consistente"
+                    " with the number of return values defined in pyfunc"
+                )
+            if len(otypes) > 1:
+                for t in otypes:
+                    if t != otypes[0]:
+                        raise NotImplementedError(
+                            "cuNumeric doesn't support variable types"
+                            " in otypes"
+                        )
+                self._output_dtype = np.dtype(otypes[0])
+
+        # FIXME
+        if excluded is not None:
+            raise NotImplementedError(
+                "excluded variables are not supported yet"
+            )
+
+        # FIXME
+        if signature is not None:
+            raise NotImplementedError(
+                "signature variable is not supported yet"
+            )
+
+    def _get_func_body(self, func: Callable[[Any], Any]) -> list[str]:
+        """Using the magic method __doc__, we KNOW the size of the docstring.
+        We then, just subtract this from the total length of the function
+        """
+        lines_to_skip = 0
+        if func.__doc__ is not None and len(func.__doc__.split("\n")) > 0:
+            lines_to_skip = len(func.__doc__.split("\n"))
+
+        lines = inspect.getsourcelines(func)[0]  # type ignore
+
+        return_lines = []
+        for i in range(lines_to_skip + 1, len(lines)):
+            return_lines.append(lines[i].rstrip())
+        return return_lines
+
+    def _get_return_argumets(self) -> list[str]:
+        """
+        Returns the list of names for return arrays/values
+        """
+        self._func_body = self._get_func_body(self._pyfunc)
+        return_names = []
+        for ln in self._func_body:
+            if "return" in ln:
+                ln = ln.replace("return", "")
+                ln = ln.replace(" ", "")
+                return_names = ln.split(",")
```

Comment:
I have just realized that this implementation doesn't support following cases:
1) there are multiple "return" statements in the UDF
2) UDF returns an output of some computation (`return a+b`)
I will try to fix it today

Reply:
I have fixed 1, and added the check for 2.